### PR TITLE
Introduce `lu!` for `UmfpackLU` to make use of its symbolic factorization

### LINKED ISSUE #665: Expose a symbolic LU factorization for sparse matrices
It would be useful to have an interface for symbolic LU factorizations with UMFPACK, similar to the CHOLMOD factorizations

### LINKED ISSUE #33323: Expose a symbolic LU factorization for sparse matrices
It would be useful to have an interface for symbolic LU factorizations with UMFPACK, similar to the CHOLMOD factorizations

### PR DESCRIPTION
**Pull request by goggle**
This PR introduces a `lu!` method for `UmfpackLU` to make use of a previously computed symbolic factorization (closes JuliaLang/LinearAlgebra.jl#665).

Here are some benchmarks:
```julia
using LinearAlgebra
using SparseArrays

nSamples = 10000
n = 40
A = sparse(sprandn(n, n, 0.1) + Matrix(10.0*I, n, n))
matrices = [sparse(SparseMatrixCSC(n, n, copy(A.colptr), copy(A.rowval), randn(nnz(A))) + Matrix(10.0*I, n, n)) for i=1:nSamples]

function naive_lu(matrices)
    for A in matrices 
        F = lu(A)
    end
end

function reuse_lu(matrices)
    F = lu(matrices[1])
    for i=2:length(matrices)
        lu!(F, matrices[i])
    end
end
```

```julia
julia> using BenchmarkTools

julia> @btime naive_lu(matrices)
  1.584 s (623885 allocations: 862.68 MiB)

julia> @btime reuse_lu(matrices)
  1.182 s (344645 allocations: 707.05 MiB)
```

### CODE REVIEW

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (dkarrasch) [inline stdlib/SuiteSparse/src/umfpack.jl]: So the fact that you're re-using the symbolic decomposition is expressed by not calling `finalizer(umfpack_free_symbolic, res)`, compared to `lu`?

File: stdlib/SuiteSparse/test/umfpack.jl
Comment (dkarrasch) [inline stdlib/SuiteSparse/test/umfpack.jl]: From where is this error thrown? I only see that you _set_ the colptrs, rowvals, and nzvals of `F` by those of `S`, but I can't seem to find where you check whether these are consistent with what you precomputed.

**Review by dkarrasch — COMMENTED** [review]
Very nice! This is outside of what I typically work with, so I only have some confusion to share with you.

**Comment by goggle** [comment]
Here are some more explanation:
The [UMFPACK QuickStart Guide](https://fossies.org/linux/SuiteSparse/UMFPACK/Doc/UMFPACK_QuickStart.pdf) explains that there are 5 primary routines for the factorization of a sparse matrix `A` in the context of `A * x = b`. Important for us are the routines `umfpack_di_symbolic` and `umfpack_di_numeric`. The routine `umfpack_di_symbolic` performs a symbolic factorization of `A`. The routine `umfpack_di_numeric` uses the information from the previously performed symbolic factorization and calculates the numeric factorization of `A`.

Prior to this PR, the calls to these two routines were coupled: Every time we want to perform a LU factorization of `A`, we had to first call `umfpack_di_symbolic` and then `umfpack_di_numeric`. This PR adds the ability to skip the call for `umfpack_di_symbolic` if we want to perform a LU factorization of a matrix `B` with the same sparsity pattern as the matrix `A`. This is done by modifying the Julia method `umfpack_numeric!(U::UmfpackLU)`, such that it does not return anymore when a numeric factorization is already stored in `U`, but instead recomputes the numeric factorization without calling `umfpack_di_symbolic` first. It makes use of the already stored symbolic factorization stored in `U` and `UMFPACK` itself handles errors (like different sparsity patterns) by returning a status code.

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (goggle) [reply stdlib/SuiteSparse/src/umfpack.jl]: @dkarrasch See my comment below, I hope this helps.

File: stdlib/SuiteSparse/test/umfpack.jl
Comment (goggle) [reply stdlib/SuiteSparse/test/umfpack.jl]: The error is thrown in line `225`. Because the status code `F.status` is not equal to `UMFPACK_OK`, `issuccess(F)` is false and the error is thrown.

**Comment by dkarrasch** [comment]
I see, so the sparsity pattern check is internal in UMFPACK.

**Comment by goggle** [comment]
@dkarrasch Exactly, yes.

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (dkarrasch) [inline stdlib/SuiteSparse/src/umfpack.jl]: I know this not done in many other places in this file, but I thought that some Julia style guide asks not to have spaces when setting keyword arguments, also in line 216.

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (dkarrasch) [inline stdlib/SuiteSparse/src/umfpack.jl]: I started being confused about the fact that we also convert in `!` methods, but that seems to be consistent with `cholesky!`, and so far nobody complained, so I guess that's ok.

**Review by dkarrasch — APPROVED** [review]
LGTM, except for the minor style comment, but that style is probably not applied elsewhere in this file, so we could as well leave it.

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (goggle) [inline stdlib/SuiteSparse/src/umfpack.jl]: These copy instructions are certainly quite expensive.
I was wondering if there is a way to prevent these?

**Comment by dkarrasch** [comment]
I'm not sure, but is there a place in the Julia docs to show the docstring of the sparse `lu!`. I couldn't find a page specifically on "sparse linear algebra". It's just when you ask the help, you get all those "sparse" docstrings, but they don't seem to appear in the julia manual.

File: stdlib/SuiteSparse/src/umfpack.jl
Comment (dkarrasch) [reply stdlib/SuiteSparse/src/umfpack.jl]: I don't see any way to prevent it. We do the same in `lu` above, and can't write into the fields of `S`.

Comment (goggle) [reply stdlib/SuiteSparse/src/umfpack.jl]: Yes, this was just a thought that came to my mind, if it was technically possible to restructure `umfpack.jl`, such that UMFPACK has direct access to the data and copying the data can be avoided. I think it's not a big deal if we perform a LU decomposition once, but it probably has a big impact for repeatedly performed LU factorizations.

Comment (dkarrasch) [reply stdlib/SuiteSparse/src/umfpack.jl]: Nevertheless, your PR is a strict improvement, because we copy as in `lu`, but save the symbolic factorization step, right? I'm trying to understand if this issue is merge-blocking, or if this is something for future development and the PR can be merged as is.

Comment (goggle) [reply stdlib/SuiteSparse/src/umfpack.jl]: Yes exactly, this PR allows the user to reuse the symbolic factorization for repeatedly performed LU decompositions.

My comment above is out-of-scope for this PR, but might be interesting to think about for future development.

**Comment by ViralBShah** [comment]
What's the status on this? Does this need another review before we merge?

**Comment by ViralBShah** [comment]
Does `lu!` mutate one of the input arguments?

**Comment by dkarrasch** [comment]
> Does `lu!` mutate one of the input arguments?

Yes, it does mutate the first argument `F::UmfpackLU` IIUC.

**Comment by ViralBShah** [comment]
In that case is this just a case of rebase and merge?

**Comment by dkarrasch** [comment]
> In that case is this just a case of rebase and merge?

I think so. I wasn't confident enough to simply merge at the time, and was hoping someone else would have a look. But the feature was asked for, so we should let people use and test it in the wild.

**Comment by ViralBShah** [comment]
The PR looks ok. It would have been nice to be able to have a more to reuse the symbolic factorization too. What would that syntax look like?

**Comment by j-fu** [comment]
Very nice to see this. This is certainly important for implicit Euler timestepping and path following for PDEs. Will try this out as soon as it becomes available and I can afford the time (unfortunately not before mid February).

**Comment by ViralBShah** [comment]
Perhaps @Chrisrackauckas and @yingboma may also want it in DiffEq.

**Comment by ChrisRackauckas** [comment]
yes I do.

**Comment by ViralBShah** [comment]
Will be out in 1.5. would be good to test it on master.

**Comment by YingboMa** [comment]
This is very nice! Thanks for your work!

```julia
julia> @btime naive_lu(matrices)
  1.584 s (623885 allocations: 862.68 MiB)

julia> @btime reuse_lu(matrices)
  1.182 s (344645 allocations: 707.05 MiB)
```
I am surprised that the improvement isn’t very dramatic. Did Julia cache something before?

**Comment by ViralBShah** [comment]
Symbolic factorization is usually only 10-20% of the factorization time. We already had the capability of reusing the numeric factorization.

**Comment by j-fu** [comment]
May be the picture changes  a bit when setting the pivot tolerance to zero (if you can afford this for your problem, certainly this cannot be the default). The hypothesis is that there would be no reordering steps from partial pivoting, making the numerical factorization relatively cheaper. Not sure though.

**Comment by YingboMa** [comment]
Thanks for the insight, I will keep that in mind when I implement the feature in JuliaDiffEq.

**Pull request merged**
